### PR TITLE
CC1310 Long Range Mode 5kbps build

### PR DESCRIPTION
**Pull request by tbriggs6**
Changes to support 5Kbps LRM  on the TI CC1310 rev B chips.

I wanted to be able to switch back and forth, so everything is conditionally compiled based on the definition of the `CC1310_LRM` variable
```
$ make TARGET=cc26x0-cc13x0 BOARD=launchpad/cc1310  CC1310_LRM=1
```
Will build the LRM version, omitting the `CC1310_LRM` will build the default 50kbps GFSK version.  To do this, I made a couple of minor changes to the file layout:

- Moved the original / default smartrf_settings.* to a subdirectory `rf-core/gfsk` 
- Added then the new 5kbps version of smartrf settings to `rf-core/lrm` 
- Adjusted the retransmission timeouts for CSMA and sicslopan 
- Adjusted the watchdog timer to deal with longer time spent waiting for packets to finish (fixed reboots under heavy load)
- Migrated the `RF_Mode` struct from smartrf-settings, and then modified `prop-mode.c` to call the function pointers to the RF patches instead of the explicit names

I ran this code with the rpl-border-router and cc26x0 web demo.  I was able to communicate for days without crashing.  Can simultaneously ping and hit the web page.  Performance is obviously 10x slower than the standard 50kbps, but it works.  I also noticed increased range.  Using the launchpad boards as reference, I'm getting about 85% of the expected range from the TI range estimator.

### CODE REVIEW

**Comment by g-oikonomou** [comment]
I've discussed this with a collaborator who did not manage to build this - there are some header files required by LRM that we don't ship in the submodule.

The reason travis passes is we do not have a test for LRM. It woule be nice to add a tets for LRM compile and to also figure out a way to ship this headers - not sure about the best way to handle the latter.

**Comment by tbriggs6** [comment]
> I've discussed this with a collaborator who did not manage to build this - there are some header files required by LRM that we don't ship in the submodule.

Right.  I forgot about the 3 newer rf patches from TI.   I just diff'ed the versions that are in the cc13xxware projects now vs. the new ones, and they are different.  Does licensing allow us to add them to the cc13xxware "git" project?  I can do a pull request on that project to add the 3 header files.
 
> The reason travis passes is we do not have a test for LRM. It woule be nice to add a tets for LRM compile and to also figure out a way to ship this headers - not sure about the best way to handle the latter.

If you accept the idea of modding the cc13xxware to add the three headers, I'll go ahead and do that.  I can also create an issue in your track for the travis test and I'll poke at it.
-

**Comment by g-oikonomou** [comment]
I am not particularly keen on adding the headers to the cc13xxware repo because them the repo would stop being an exact git version of the respective xxware release.

I would rather ship them under `rf-core/lrm` and only include them when using LRM, perhaps via inclusion of an umbrella header that we place in both lrm and gfsk subdirs (something like `rf-patches.h` maybe?). 

> Does licensing allow us to add them to the cc13xxware "git" project? 

Do they have a copyright header? What does it say?

Other than that, I can see that there are some unrelated cosmetic changes in the diff. More in the review.

File: arch/cpu/cc26x0-cc13x0/rf-core/prop-mode.c [hunk 1]
```diff
@@ -67,18 +67,19 @@
...
 /*---------------------------------------------------------------------------*/
-#include "rf-core/smartrf-settings.h"
+
 /*---------------------------------------------------------------------------*/
 #include <stdint.h>
 #include <string.h>
 #include <stdio.h>
 #include <stdbool.h>
 /*---------------------------------------------------------------------------*/
-#define DEBUG 0
```
Comment (g-oikonomou) [inline arch/cpu/cc26x0-cc13x0/rf-core/prop-mode.c]: Please remove from the diff

File: arch/cpu/cc26x0-cc13x0/rf-core/prop-mode.c [hunk 2]
```diff
@@ -381,6 +382,8 @@ set_tx_power(radio_value_t power)
     }
   }
 }
+
+
```
Comment (g-oikonomou) [inline arch/cpu/cc26x0-cc13x0/rf-core/prop-mode.c]: Please remove from the diff

File: arch/cpu/cc26x0-cc13x0/rf-core/prop-mode.c [hunk 3]
```diff
@@ -396,12 +399,15 @@ prop_div_radio_setup(void)
   /* Update to the correct TX power setting */
   smartrf_settings_cmd_prop_radio_div_setup.txPower = tx_power_current->tx_power;
 
+
```
Comment (g-oikonomou) [inline arch/cpu/cc26x0-cc13x0/rf-core/prop-mode.c]: Please remove from the diff

File: arch/cpu/cc26x0-cc13x0/rf-core/prop-mode.c [hunk 4]
```diff
@@ -396,12 +399,15 @@ prop_div_radio_setup(void)
...
   /* Adjust RF Front End and Bias based on the board */
   smartrf_settings_cmd_prop_radio_div_setup.config.frontEndMode =
     RF_CORE_PROP_FRONT_END_MODE;
   smartrf_settings_cmd_prop_radio_div_setup.config.biasMode =
     RF_CORE_PROP_BIAS_MODE;
 
+
+
```
Comment (g-oikonomou) [inline arch/cpu/cc26x0-cc13x0/rf-core/prop-mode.c]: Please remove from the diff

File: arch/cpu/cc26x0-cc13x0/rf-core/prop-mode.c [hunk 5]
```diff
@@ -457,7 +463,6 @@ rf_cmd_prop_rx()
            cmd_status, cmd_rx_adv->status);
     return RF_CORE_CMD_ERROR;
   }
-
```
Comment (g-oikonomou) [inline arch/cpu/cc26x0-cc13x0/rf-core/prop-mode.c]: Please remove from the diff

File: arch/cpu/cc26x0-cc13x0/rf-core/prop-mode.c [hunk 6]
```diff
@@ -812,7 +817,6 @@ release_data_entry(void)
   critical_exit(interrupt_status);
 
 }
-/*---------------------------------------------------------------------------*/
```
Comment (g-oikonomou) [inline arch/cpu/cc26x0-cc13x0/rf-core/prop-mode.c]: Please remove from the diff

File: arch/cpu/cc26x0-cc13x0/rf-core/prop-mode.c [hunk 7]
```diff
@@ -845,7 +849,7 @@ read_frame(void *buf, unsigned short buf_len)
   len = (*(uint16_t *)data_ptr);
 
   if(len <= RX_BUF_METADATA_SIZE) {
-    PRINTF("RF: too short!");
+
```
Comment (g-oikonomou) [inline arch/cpu/cc26x0-cc13x0/rf-core/prop-mode.c]: Please remove from the diff

File: arch/cpu/cc26x0-cc13x0/rf-core/prop-mode.c [hunk 8]
```diff
@@ -1057,10 +1061,17 @@ on(void)
     rf_core_set_modesel();
 
     /* Apply patches to radio core */
-    rf_patch_cpe_genfsk();
+    if (RF_prop.cpePatchFxn != 0)
+        RF_prop.cpePatchFxn( );
+
+    if (RF_prop.mcePatchFxn != 0)
+        RF_prop.mcePatchFxn( );
+
```
Comment (g-oikonomou) [inline arch/cpu/cc26x0-cc13x0/rf-core/prop-mode.c]: We'll need to do something more clever here and somehow use a single call/macro that can select and apply the correct patches.

File: arch/cpu/cc26x0-cc13x0/rf-core/prop-mode.c [hunk 9]
```diff
@@ -1057,10 +1061,17 @@ on(void)
...
     while(!HWREG(RFC_DBELL_BASE + RFC_DBELL_O_RFACKIFG));
     HWREG(RFC_DBELL_BASE + RFC_DBELL_O_RFACKIFG) = 0;
-    rf_patch_rfe_genfsk();
+
+    if (RF_prop.rfePatchFxn != 0)
+        RF_prop.rfePatchFxn( );
```
Comment (g-oikonomou) [inline arch/cpu/cc26x0-cc13x0/rf-core/prop-mode.c]: Ditto

File: arch/cpu/cc26x0-cc13x0/rf-core/prop-mode.h [hunk 10]
```diff
@@ -47,11 +47,15 @@
...
+#include "smartrf_fxn.h"
+
 /*---------------------------------------------------------------------------*/
 typedef struct prop_mode_tx_power_config {
   radio_value_t dbm;
   uint16_t tx_power; /* Value for the PROP_DIV_RADIO_SETUP.txPower field */
 } prop_mode_tx_power_config_t;
+
+
 /*---------------------------------------------------------------------------*/
```
Comment (g-oikonomou) [inline arch/cpu/cc26x0-cc13x0/rf-core/prop-mode.h]: Please remove from the diff

File: arch/cpu/cc26x0-cc13x0/rf-core/smartrf_fxn.h [hunk 11]
```diff
@@ -0,0 +1,14 @@
...
+typedef struct {
+    int rfMode;
+    void (*cpePatchFxn)(void);
+    void (*mcePatchFxn)(void);
+    void (*rfePatchFxn)(void);
+
+} RF_Mode;
+
+#endif
+
```
Comment (g-oikonomou) [inline arch/cpu/cc26x0-cc13x0/rf-core/smartrf_fxn.h]: Is this file coming from SmartRF studio or is it from the SDK? Need to figure out the license on it

Comment (moskito-s) [reply arch/cpu/cc26x0-cc13x0/rf-core/smartrf_fxn.h]: I think it comes from the simplelink SDK 1.6

**Comment by moskito-s** [comment]
After including the header files I can get it to run. My cc1310s are able to communicate. I was however unable to get a RPL-Network running. I tried MRHOF and OF0. It seems that the slow transfer rate causes some timing issues

**Comment by rajeev1986** [comment]
> After including the header files I can get it to run. My cc1310s are able to communicate. I was however unable to get a RPL-Network running. I tried MRHOF and OF0. It seems that the slow transfer rate causes some timing issues

The issue is that correct CSMA timing is not picked up when compiling `CC1310_LRM=1` directly.
You need to include the following  FLAGS in your makefile to pick up the correct timings.
```
ifdef CC1310_LRM
 CFLAGS += -DCC1310_LRM=1
endif
```

**Comment by moskito-s** [comment]
> 
> The issue is that correct CSMA timing is not picked up when compiling `CC1310_LRM=1` directly.
> You need to include the following FLAGS in your makefile to pick up the correct timings.
> 
> ```
> ifdef CC1310_LRM
>  CFLAGS += -DCC1310_LRM=1
> endif
> ```

Thank you for the hint. It runs now.

> I have created a simple example [here](https://github.com/rajeev1986/contikiNG-tutorials/tree/master/rssi-csma-lrm) to demonstrate the LRM running atop CSMA and RPL-lite. Set your Contiki-ng link (`CONTIKI = ../../..`) correctly using @tbriggs6's `cc1310-lrm`repo, compile `make CC1310_LRM=1` and should work. I have tested it using two CC1310 Launchpads @915MHz band.

Thank you for the example. I can get it to run on my desk but as soon as I do distance testing, the rf-core seems to malfunction:

After the sender and receiver loose connection it seems that `rf_is_on()` inside of prop-mode.c always return false. Even if the boards get closer to each other again. As a result `off()` gets called and the boards can not communicate anymore: The DIS Messages don't reach the other board. This happens to the sender or receiver. Which one fails first seems random.
Calling `get_value(RADIO_PARAM_POWER_MODE)` also returns `RADIO_POWER_MODE_OFF`

Steps to reproduce
1)Reduce TX Power to -10  (I did this in order to reproduce the error without needing to travel far)
2)Start the connection with the boards close to each other
3)Leave with the sending board out of the reach of the receiving board
4)wait a few minutes
5)come back 

I also used `#define DOT_15_4G_CONF_FREQUENCY_BAND_ID DOT_15_4G_FREQUENCY_BAND_863` since I live in europe.

The last successful udp communication gave me the following log:

```
[INFO: Frame 15.4] In:  1 0012.4b00.0e09.6562 0012.4b00.13a4.8343 21 26 (47)
[INFO: CSMA      ] received packet from 0012.4b00.0e09.6562, seqno 194, len 26
[DBG : 6LoWPAN   ] uncompression: IPHC dispatch
[DBG : 6LoWPAN   ] uncompression: CID flag set - increase header with one
[DBG : 6LoWPAN   ] uncompression: address 8 0fd00::212:4b00:e09:6562
[DBG : 6LoWPAN   ] uncompression: address 8 0fd00::212:4b00:13a4:8343
[DBG : 6LoWPAN   ] uncompression: found ext header id: 0 next: 0 len: 6
[DBG : 6LoWPAN   ] uncompression: 0 len: 6 exhdrlen: 0 (calc: 8)
[DBG : 6LoWPAN   ] uncompression: incoming header value: 240
[DBG : 6LoWPAN   ] uncompression: UDP ports (ptr+5): 223d, 162e
[DBG : 6LoWPAN   ] uncompression: checksum included
[DBG : 6LoWPAN   ] uncompression: UDP length: 16 (ext: 8) ip_len: 0 udp_len: 16
[DBG : 6LoWPAN   ] uncompression: IP payload length: 24. 26 - 18 + 56 - 40
[INFO: 6LoWPAN   ] input: received IPv6 packet with len 64
[DBG : 6LoWPAN   ] uncompression: after (24):6000000000180040fd0000000000000002124b000e096562fd0000000000000002124b0013a48343110063040000011b223d162e0010d2c348656c6c6f203234
[INFO: TCP/IP    ] input: received 64 bytes
[INFO: IPv6      ] packet received from fd00::212:4b00:e09:6562 to fd00::212:4b00:13a4:8343
[DBG : IPv6      ] Processing RPL option
[INFO: RPL       ] ext hdr: packet from fd00::212:4b00:e09:6562 to fd00::212:4b00:13a4:8343 going up, sender closer 0 (283 < 128), rank error 0, loop detected 0
[DBG : IPv6      ] Processing hbh header
[DBG : IPv6      ] Processing RPL option
[INFO: RPL       ] ext hdr: packet from fd00::212:4b00:e09:6562 to fd00::212:4b00:13a4:8343 going up, sender closer 0 (283 < 128), rank error 0, loop detected 0
[DBG : IPv6      ] Removing IPv6 extension headers (extlen: 8, uiplen: 64)
[INFO: IPv6      ] Receiving UDP packet
[DBG : IPv6      ] Upper layer checksum len: 16 from: 40
[DBG : IPv6      ] In udp_found
[INFO: App       ] Received request 'Hello 24' from fd00::212:4b00:e09:6562
[INFO: App       ] Last RSSI -102

[INFO: App       ] Sending response.
[DBG : IPv6      ] In udp_send
[DBG : IPv6      ] Upper layer checksum len: 16 from: 40
[INFO: IPv6      ] Sending packet with length 56 (16)
[INFO: RPL       ] SRH creating source routing header with destination fd00::212:4b00:e09:6562 
[INFO: RPL       ] SRH node not found, skip SRH insertion
[INFO: TCP/IP    ] output: processing 56 bytes packet from fd00::212:4b00:13a4:8343 to fd00::212:4b00:e09:6562
[DBG : RPL       ] no SRH found
[ERR : TCP/IP    ] output: destination off-link and no default route
[DBG : IPv6      ] In udp_send
[INFO: RPL       ] nbr: own state, addr fd00::212:4b00:13a4:8343, DAG state: reachable, MOP 1 OCP 1 rank 128 max-rank 65535, dioint 17, nbr count 0 (Periodic)
[INFO: RPL       ] nbr: end of list
[INFO: RPL       ] No routing links
[INFO: Frame 15.4] In:  1 0012.4b00.0e09.6562 0000.0000.0000.0000 15 80 (95)
[INFO: CSMA      ] received packet from 0012.4b00.0e09.6562, seqno 195, len 80
[DBG : 6LoWPAN   ] uncompression: IPHC dispatch
[DBG : 6LoWPAN   ] uncompression: next header inline: 254
[DBG : 6LoWPAN   ] uncompression: address 0 161a9b:160:1902:f001:1808:f000:167d:efef
[DBG : 6LoWPAN   ] uncompression: IP payload length: 61. 80 - 19 + 40 - 40
[INFO: 6LoWPAN   ] input: received IPv6 packet with len 101
[DBG : 6LoWPAN   ] uncompression: after (61):60000000003dfe401a9b01601902f0011808f000167defeffd0000000000000002124b000e0965624a1031c7aede7dcd0813a48343040e2cb667c471605f8e9ede214c316fd542379b5b949e1a7d5eb64d364e8850ead78b6568d085552947bf3354515ebb
[INFO: TCP/IP    ] input: received 101 bytes
[ERR : IPv6      ] invalid extension header chain
[INFO: RPL       ] nbr: own state, addr fd00::212:4b00:13a4:8343, DAG state: reachable, MOP 1 OCP 1 rank 128 max-rank 65535, dioint 18, nbr count 0 (Periodic)
[INFO: RPL       ] nbr: end of list
[INFO: RPL       ] No routing links
[INFO: RPL       ] nbr: own state, addr fd00::212:4b00:13a4:8343, DAG state: reachable, MOP 1 OCP 1 rank 128 max-rank 65535, dioint 18, nbr count 0 (Periodic)
[INFO: RPL       ] nbr: end of list
[INFO: RPL       ] No routing links
[INFO: RPL       ] sending a multicast-DIO with rank 128 to ff02::1a
[DBG : IPv6      ] Upper layer checksum len: 76 from: 40
[INFO: ICMPv6    ] Sending ICMPv6 packet to ff02::1a, type 155, code 1, len 72
[INFO: TCP/IP    ] output: sending to (NULL LL addr)
[INFO: 6LoWPAN   ] output: sending IPv6 packet with len 116
[DBG : 6LoWPAN   ] compression: before (76): 60000000004c3a40fe8000000000000002124b0013a48343ff02000000000000000000000000001a9b013d3500f0008008f00000fd0000000000000002124b0013a48343040e00080c00040000800001001e003c081e4040ffffffffffffffff00000000fd000000000000000000000000000000
[DBG : 6LoWPAN   ] compression: first header: 58
[DBG : 6LoWPAN   ] compression: last header could is not compressed: 58
[DBG : 6LoWPAN   ] compression: after (4): 7a3b3a1a
[INFO: 6LoWPAN   ] output: header len 40 -> 4, total len 116 -> 80, MAC max payload 110, frag_needed 0
[INFO: CSMA      ] sending to 0000.0000.0000.0000, len 80, seqno 21, queue length 1, free packets 7
[DBG : CSMA      ] scheduling transmission in 0 ticks, NB=0, BE=3
[INFO: CSMA      ] preparing packet for 0000.0000.0000.0000, seqno 21, tx 0, queue 1
[INFO: Frame 15.4] Out:  1 ffff.0000.0000.0000 15 80 (95)
[INFO: CSMA      ] tx to 0000.0000.0000.0000, seqno 21, status 0, tx 0, coll 0
[INFO: CSMA      ] packet sent to 0000.0000.0000.0000, seqno 21, status 0, tx 1, coll 0
[DBG : CSMA      ] free_queued_packet, queue length 0, free packets 8
[INFO: RPL       ] nbr: own state, addr fd00::212:4b00:13a4:8343, DAG state: reachable, MOP 1 OCP 1 rank 128 max-rank 65535, dioint 18, nbr count 0 (Periodic)
[INFO: RPL       ] nbr: end of list
[INFO: RPL       ] No routing links

**Comment by moskito-s** [comment]
@rajeev1986 
CSMA under the same test conditions and without the LRM enabled worked just fine. 


> ```
> #if CC1310_LRM
> /* in LRM mode set the bit rate to 5kbps */
> #define CC13XX_RADIO_BIT_RATE 5000
> #else
> /* default bit rate is 50kbps */
> #define CC13XX_RADIO_BIT_RATE 50000
> #endif
> ```
ll try that. 
On another note: since the LRM uses a higher sensitivity setting. Shouldn't `PROP_MODE_CONF_RSSI_THRESHOLD`
also be decreased to -120dBm?

**Comment by g-oikonomou** [comment]
I do not think this is mature enough for inclusion here, perhaps best kept in a branch somewhere. Closing, for now at least.

**Pull request closed without merging**